lib/puppet/functions/reboot/wait.rb

```diff
@@ -0,0 +1,70 @@
+# Waits for nodes to reboot when executed from within a plan
+#
+# This function assumes that the nodes has just been told to reboot and
+# therefore waits for it to disconnect and reconnect again.
+#
+# This has no valid use outside plans!
+Puppet::Functions.create_function(:'reboot::wait') do
+  # @param targets A TargetSpec containing all targets to wait for
+  # @param params Extra parameters defined as a hash, valid keys are:
+  #   disconnect_wait, reconnect_wait and retry_interval. All values should be
+  #   integers and represent seconds.
+  # @example Wait for some very slow nodes to reboot.
+  #   reboot::wait($nodes, { 'disconnect_wait' => 120, 'reconnect_wait' => 600 })
+  dispatch :wait do
+    required_param 'Tuple', :targets
+    optional_param 'Hash', :params
+  end
+
+  def wait(targets, params = { disconnect_wait: 20, reconnect_wait: 120, retry_interval: 1 })
+    threads = []
+
+    targets.each do |target|
+      # We need to thread this so that we can check many nodes at once, it is
+      # possible that this could cause performance issues but only at very
+      # large scales and if you've just triggered 5000 nodes to reboot at once
+      # you have bigger problems...
+      threads << Thread.new do
+        begin
+          # If the target is connected in the beginning, wait for it to disconnect,
+          # then wait for it to come back. Some server may take many minutes to shut
+          # down.
+          wait_until(params[:disconnect_wait], params[:retry_interval]) { !connected?(target) }
+
+          # Once the target has disconnected, wait for it to come back
+          wait_until(params[:reconnect_wait], params[:retry_interval]) { connected?(target) }
+        rescue StandardError
+          raise "Timed out waiting for #{target.name} to reboot"
+        end
+      end
+    end
+
+    threads.each(&:join)
+  end
+
+  def connected?(target)
+    executor  = Puppet.lookup(:bolt_executor) { nil }
+    transport = executor.transports[target.protocol].value
+
+    case transport
```

Comment:
This is interesting. It could break with future Bolt releases.

When we've discussed this before, we talked about making it a plan that would make use of run_command/run_task with `_catch_errors => true` to detect when the connection was available again. Using the inventory directly is probably more reliable for the Orch transport, but also more invasive.

Reply:
Yeah it definitely could, I especially dislike having to use `instance_variable_get` but that's the only way to do it. I can see why plans would be good as they don't really have to do differentiate between the different transports but they also require some kind of "dummy" task to run and would be pretty ugly in the logs. I've now added tests to ensure that the things we rely on in Bolt are still there so we should at least get failures if things change.